### DISCUSSION TITLE
Coupling Two System Variables on the Boundaries

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#19365] [Q&A General] [answered]
Hi
I am trying to simulate a coupled system where the system variables are only coupled on the boundaries. For example,
the two systems in a simpler form would be as follows:
system 1:
![image](https://user-images.githubusercontent.com/88537645/141350636-25f8d512-362e-4f5a-93d2-d1987d924811.png)

and system 2:
![image](https://user-images.githubusercontent.com/88537645/141350704-7bce7476-2dba-41bf-8899-3a51cf34d6b2.png)
As a trial, I used one input file to define both PDEs and used `CoupledVarNeumannBC` to model the robin BCs of coupled variables, but it does not show any coupling. 

Any thoughts in this regard?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Do both systems live on the same cells or are they on both sides of the boundary?

Guillaume

**Reply by aaelmeli** [reply]
> Hello
> 
> Do both systems live on the same cells or are they on both sides of the boundary?
> 
> Guillaume

Yes, both systems live on the same cells.

**Reply by GiudGiud** [reply]
Should have worked afaik then. 
Can you share your input file?

**Reply by aaelmeli** [reply]
> Should have worked afaik then. Can you share your input file?

The following is the input file for the simplest case, a1D problem:
```
> [Mesh]
> 
>         type = GeneratedMesh
>         dim = 1
>         xmin = 0
>         xmax = 1.0
>         nx = 1000
> 
> []

> [Variables]
> 
>     [ureal]
>         order = FIRST
>         family = LAGRANGE
>     []
>     [uimag]
>         order = FIRST
>         family = LAGRANGE
>     []
> 
> []

> [Kernels]
> 
>     [diffusion_real]
>         type = MatDiffusion
>         variable = ureal
>         diffusivity = _diff_
>     []
> 
>     [diffusion_imag]
>         type = MatDiffusion
>         variable = uimag
>         diffusivity = _diff_
>     []
>
>     [reaction_real]
>         type = Reaction
>         variable = ureal
>         rate = -10000
>     []
> 
>     [reaction_imag]
>         type = Reaction
>         variable = uimag
>         rate = -10000
>     []
> []
> 
> 
> [Materials]
> 
>     [mat]
>     type = GenericConstantMaterial
>     prop_names = '_diff_'
>     prop_values = '1'
>     []
> 
> []


> [BCs]
> 
>     [BC_left_real]
>         type = CoupledVarNeumannBC
>         variable = ureal
>         boundary = 'left'
>         v = uimag
>         coef=-100
>     []
> 
>      [BC_left_imag]
>         type = CoupledVarNeumannBC
>         variable = uimag
>         boundary = 'left'
>         v = ureal
>         coef=100
>     []
> 
>     [BC_right_real]
>         type = DirichletBC
>         variable = ureal
>         boundary = 'right'
>         value = 1
>     []
> 
>     [BC_right_imag]
>         type = DirichletBC
>         variable = uimag
>         boundary = 'right'
>         value = 0
>     []
> 
> []
> 
> 
>    [Executioner]
> 
>       execute_on = 'initial timestep_end'
>        type = Steady
>        solve_type = LINEAR
> 
>   []
> 
>   [Outputs]
> 
>     [exodus]
>         type = Exodus
>         execute_on = final
>     [ ]
> []
```

**Reply by GiudGiud** [reply]
You are using solve_type = LINEAR. I dont think this is a linear problem anymore with the coupling. If you multiply u by 2, 2*v is not a solution. 

Just remove that line and I got a believable answer.
Note that execute_on is not a parameter to Executioner. You should have been getting a warning about this.

**Reply by aaelmeli** [reply]
> You are using solve_type = LINEAR. I dont think this is a linear problem anymore with the coupling. If you multiply u by 2, 2*v is not a solution.
> 
> Just remove that line and I got a believable answer. Note that execute_on is not a parameter to Executioner. You should have been getting a warning about this.

Yes, that was the problem. Thank you. It works now.

**Reply by aaelmeli** [reply]
Hi @GiudGiud, just a follow-up question, although using solver_type = NEWTON results in variables coupling, we have noticed that the system of equation is a linear system in the sense that, if the force (usually point sources) is multiplied by 2, both `u `and `v` variables will be multiplied by 2. 
Basically, it is a decomposed version of the following Helmholtz equation (which is a linear system)
![image](https://user-images.githubusercontent.com/88537645/142896973-1f0e81f7-a56c-4248-99e5-7f7700efcbd6.png)
discussed in this forum #19259. And there is no nonlinear transformation that has been applied when doing that decomposition.

So, I was hoping if we can assemble the entire system as `Ax=b` , that is the coupling effect done through `CoupledVarNeumannBC` goes directly to the `A` and be able to solve the system using direct solvers. Any thought on this?
Thank you.

**Reply by GiudGiud** [reply]
That is not true with the DirichletBC of 1 for ureal on the right boundary. If you change that one, then it s more likely to be linear. The NeumannBCs and the 0 DirchletBC certainly allow for linear transformations of the variables.

**Comment by aaelmeli** [comment]
> That is not true with the DirichletBC of 1 for ureal on the right boundary. If you change that one, then it s more likely to be linear. The NeumannBCs and the 0 DirchletBC certainly allow for linear transformations of the variables.

Yes, you are right. And that what I have tried, please see the following input file. Using this input, I could not get couplding between the variables.
```
[Mesh]
        type = GeneratedMesh
        dim = 1
        xmin = 0
        xmax = 1.0
        nx = 1000
[]

[Variables]
    [ureal]
        order = FIRST
        family = LAGRANGE
    []
    [uimag]
        order = FIRST
        family = LAGRANGE
    []
[]

[Kernels]
    [diffusion_real]
        type = MatDiffusion
        variable = ureal
        diffusivity = _diff_
    []
    [diffusion_imag]
        type = MatDiffusion
        variable = uimag
        diffusivity = _diff_
    []
    [reaction_real]
        type = Reaction
        variable = ureal
        rate = -10000
    []
    [reaction_imag]
        type = Reaction
        variable = uimag
        rate = -10000
    []
[]

[DiracKernels]
    [source] 
        type = ConstantPointSource
        point = '1 0.0 0.0'
        variable = ureal 
        value = 1
    []
[]

[Materials]
  [mat]
    type = GenericConstantMaterial
    prop_names = '_diff_'
    prop_values = '1'
  []
[]

[BCs]

   [BC_left_real]
        type = CoupledVarNeumannBC
        variable = ureal
        boundary = 'left'
        v = uimag
        coef=-100
    []
       
    [BC_left_imag]
        type = CoupledVarNeumannBC
        variable = uimag
        boundary = 'left'
        v = ureal
        coef=100
    []
    [BC_right_imag]
        type = DirichletBC
        variable = uimag
        boundary = 'right'
        value = 0
    []
[]


[Executioner]
  type = Steady
  solve_type=LINEAR
[]

[Outputs]
    [exodus]
        type = Exodus
        execute_on = final
    []
[]
```

**Reply by GiudGiud** [reply]
The problem is likely underspecified. For a diffusion (Cauchy) problem, you need 2 BCs per variable.
Maybe it works with a natural BC but I m not sure

**Reply by GiudGiud** [reply]
or do you mean that the solution looks uncoupled?

**Comment by aaelmeli** [comment]
> or do you mean that the solution looks uncoupled?

Yes, the solution looks uncoupled whenever using `solver_type=Linear`. So I do not think that the problem is underspecified.

**Reply by GiudGiud** [reply]
the DiracKernel source is not a linear term?
Are is this the term that you are multiplying?

**Reply by aaelmeli** [reply]
Yes, this is how I apply point sources. Are there other methods you suggest to apply point sources that reserve the linearity of the problem?

**Reply by GiudGiud** [reply]
I think a point source term can be linear in those equations if it is proportional to u or v. If it s a constant then that s not linear

**Reply by aaelmeli** [reply]
Yes, the point source is linearly proportional to `u` and `v`. If I multiplied the point source by 2, the response (`u` and `v`) will be multiplied by 2. I have just verified that on moose using the former input file but I used a nonlinear solver to get the coupling.

**Reply by GiudGiud** [reply]
Ok that is what I mean. The response is linear in the value of the point source

**Reply by aaelmeli** [reply]
OK, so, given the problem is linear and the source is linear w.r.t the variables, we should expect to get the coupling when using solver_type=LINEAR as you mentioned before, which is not the case now. Do you suggest a solution? sorry for reiterating this.

**Reply by GiudGiud** [reply]
Yes.
Not sure about a solution. i would dig into the manuals to understand what LINEAR is doing and why the update from one variable the other is not happening

**Reply by aaelmeli** [reply]
Upon checking the`CoupledVarNeumannBC` , I found that no Jacobian method is implemented in that source file. So, I was wondering if that may cause that issue?
![image](https://user-images.githubusercontent.com/88537645/142928324-b7c7b0b4-e663-4898-bd4d-c75687e1cef9.png)

**Reply by GiudGiud** [reply]
seems like it should have some off-diagonal terms in the jacobian
yes that can be it. The non linear solve has iterations which will make up for missing jacobian terms (not always, sometimes those make the solve diverge).

**Comment by aaelmeli** [comment]
> I think a point source term can be linear in those equations if it is proportional to u or v. If its a constant then that s not linear

Sorry, I just notice the last part of your comment, I do not understand why if the Dirac source is constant, then it would not linear problem?

**Comment by aaelmeli** [comment]
> seems like it should have some off-diagonal terms in the jacobian yes that can be it. The nonlinear solve has iterations which will make up for missing jacobian terms (not always, sometimes those make the solve diverge).

As confirmation of what you said and based on some experiments I have done on moose, yes, the nonlinear solver always diverges for 2D problems of the same coupled PDEs. Which obviously caused by the missing off-diagonal jacobian information. 

So, The question now is, Is there a way that one can implement a function to calculate these off-diagonal terms?

**Reply by GiudGiud** [reply]
For sure. I think if they exist, and my feeling without having derived them is that they should, they should be added to CoupledVarNeumannBC.

Ok let s figure this one out then let s look at 2D problems

**Reply by aaelmeli** [reply]
Sure, I just posted this as a separate discussion. I posted it almost at the same time when I received your comment. Please feel free if you prefer to keep the discussion in this forum.
